libs/pika/async_mpi/src/mpi_polling.cpp

```diff
@@ -744,77 +710,81 @@ namespace pika::mpi::experimental {
     // -----------------------------------------------------------------
     bool pool_exists() { return detail::pool_exists_; }
 
+    void register_polling()
+    {
+        auto mode = get_completion_mode();
+        // enable pika polling on the mpi pool if the handling mode needs it
+        if (detail::get_handler_mode(mode) != detail::handler_mode::yield_while)
+        {
+            PIKA_DETAIL_DP(detail::mpi_debug<1>,
+                debug(str<>("enabling polling"), get_pool_name(), detail::mode_string(mode)));
+            detail::register_polling(pika::resource::get_thread_pool(get_pool_name()));
+        }
+    }
+
     // -------------------------------------------------------------
     // initialize the pika::mpi background request handler
     // All ranks should call this function,
     // but only one thread per rank needs to do so
-    void init(bool init_mpi, bool init_errorhandler)
+    void init(bool init_mpi, bool init_errorhandler, pool_create_mode pool_mode)
     {
+        // don't allow polling code to run until init has completed
+        std::lock_guard<detail::mutex_type> lk(detail::mpi_data_.polling_vector_mtx_);
+
+        // --------------------------------------
+        // fetch completion handling and polling mode/flags
+        auto mode = get_completion_mode();
+
+        // --------------------------------------
+        // the user has asked us to call mpi_init
         if (init_mpi)
         {
-            int required = MPI_THREAD_MULTIPLE;
-            int minimal = MPI_THREAD_FUNNELED;
-            int provided;
-            pika::util::mpi_environment::init(nullptr, nullptr, required, minimal, provided);
-            if (provided < MPI_THREAD_FUNNELED)
+            int provided, required = MPI_THREAD_MULTIPLE;
+            if ((mode & detail::to_underlying(detail::handler_mode::use_pool)) &&
+                !(mode & detail::to_underlying(detail::handler_mode::request_inline)) /*&&
+                !(mode & detail::to_underlying(detail::handler_mode::completion_inline))*/)
             {
-                PIKA_DETAIL_DP(detail::mpi_debug<5>,
-                    error(str<>("pika::mpi::experimental::init"), "init failed"));
-                PIKA_THROW_EXCEPTION(pika::error::invalid_status, "pika::mpi::experimental::init",
-                    "the MPI installation doesn't allow multiple threads");
+                // when a dedicated pool is used (1 thread), and all requests are transferred to the pool, we can use MPI_THREAD_SINGLE
+                // std::int32_t ranks = pika::detail::get_env_var_as<std::int32_t>("OMPI_WORLD_SIZE", -1);
+                // if the mpi thread pool is disabled, then we can't use single threaded
+                if (pool_mode == pool_create_mode::pika_decides)
+                {
+                    // required = MPI_THREAD_SINGLE;
+                    // PIKA_DETAIL_DP(detail::mpi_debug<0>, debug(str<>("MPI_THREAD_SINGLE")));
```

Comment:
Just checking: how much of this is intentionally commented out?

Reply:
See https://app.circleci.com/pipelines/github/pika-org/pika/10746/workflows/fd4bcefe-047c-4d49-9cd8-ee810cd66322/jobs/110842?invite=true#step-102-21.

Reply:
I should have deleted that section. It is harmless left in, but it is part of future optimizations to switch to init thread SINGLE if possible.